File: app/src/main/java/org/wikipedia/views/DrawableItemDecoration.java
package org.wikipedia.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import org.wikipedia.util.ResourceUtil;

// todo: replace with DividerItemDecoration once it supports headers and footers
public class DrawableItemDecoration extends RecyclerView.ItemDecoration {
    @NonNull private final Drawable drawable;

    public DrawableItemDecoration(@NonNull Context context, @AttrRes int id) {
        this.drawable = ContextCompat.getDrawable(context,
                ResourceUtil.getThemedAttributeId(context, id));
    }

    @Override public void getItemOffsets(Rect outRect, View view, RecyclerView parent,
                                         RecyclerView.State state) {
        super.getItemOffsets(outRect, view, parent, state);
        outRect.top = drawable.getIntrinsicHeight();
        if (parent.getChildAdapterPosition(view) == state.getItemCount() - 1) {
            outRect.bottom = drawable.getIntrinsicHeight();
        }
    }

    @Override public void onDraw(Canvas canvas, @NonNull RecyclerView parent,
                                 RecyclerView.State state) {
        super.onDraw(canvas, parent, state);
        if (parent.getChildCount() == 0) {
            return;
        }

        int end = parent.getChildCount() - 1;
        for (int i = 0; i < end; ++i) {
            draw(canvas, bounds(parent, parent.getChildAt(i), true));
        }
        draw(canvas, bounds(parent, parent.getChildAt(end), true));
        draw(canvas, bounds(parent, parent.getChildAt(end), false));
    }

    private Rect bounds(@NonNull RecyclerView parent, @NonNull View child, boolean top) {
        RecyclerView.LayoutManager layoutManager = parent.getLayoutManager();

        Rect bounds = new Rect();
        bounds.right = parent.getWidth() - parent.getPaddingRight();
        bounds.left = parent.getPaddingLeft();
        int height = drawable.getIntrinsicHeight();
        bounds.top = top
                ? layoutManager.getDecoratedTop(child)
                : layoutManager.getDecoratedBottom(child) - height;
        bounds.bottom = bounds.top + height;

        return bounds;
    }

    private void draw(Canvas canvas, @NonNull Rect bounds) {
        drawable.setBounds(bounds);
        drawable.draw(canvas);
    }
}

File: app/src/main/java/org/wikipedia/views/DrawableItemDecoration.kt
package org.wikipedia.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.View
import androidx.annotation.AttrRes
import androidx.appcompat.content.res.AppCompatResources
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import org.wikipedia.util.ResourceUtil

// todo: replace with DividerItemDecoration once it supports headers and footers
class DrawableItemDecoration @JvmOverloads constructor(context: Context, @AttrRes id: Int,
                                                       private val drawStart: Boolean = false,
                                                       private val drawEnd: Boolean = true,
                                                       private val skipSearchBar: Boolean = false) : ItemDecoration() {

    private val drawable: Drawable = AppCompatResources.getDrawable(context, ResourceUtil.getThemedAttributeId(context, id))!!

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        super.getItemOffsets(outRect, view, parent, state)
        outRect.top = drawable.intrinsicHeight
        if (parent.getChildAdapterPosition(view) == state.itemCount - 1) {
            outRect.bottom = drawable.intrinsicHeight
        }
    }

    override fun onDraw(canvas: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDraw(canvas, parent, state)
        if (parent.childCount == 0) {
            return
        }

//        val startingPosition = if (parent.getChildAt(0).id == R.id.search_container && skipSearchBar) 1 else 0

        val end = parent.childCount - 1
        for (i in (if (drawStart) 0 else 0 + 1) until end) {
            draw(canvas, bounds(parent, parent.getChildAt(i), true))
        }
        if (drawStart || parent.childCount > 0 + 1) {
            draw(canvas, bounds(parent, parent.getChildAt(end), true))
        }
        if (drawEnd) {
            draw(canvas, bounds(parent, parent.getChildAt(end), false))
        }
    }

    private fun bounds(parent: RecyclerView, child: View, top: Boolean): Rect {
        val layoutManager = parent.layoutManager
        val bounds = Rect()
        bounds.right = parent.width - parent.paddingRight
        bounds.left = parent.paddingLeft
        val height = drawable.intrinsicHeight
        bounds.top = if (top) layoutManager!!.getDecoratedTop(child) else layoutManager!!.getDecoratedBottom(child) - height
        bounds.bottom = bounds.top + height
        return bounds
    }

    private fun draw(canvas: Canvas, bounds: Rect) {
        drawable.bounds = bounds
        drawable.draw(canvas)
    }
}
